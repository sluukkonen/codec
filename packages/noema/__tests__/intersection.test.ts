import { fromJson, intersection, number, object } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

const ab = intersection(object({ a: number }), object({ b: number }))

test("should parse valid intersections", () => {
  expectParseSuccess(ab, { a: 1, b: 1 })
})

test("should fail to parse non-objects", () => {
  expectParseFailure(ab, null)
})

test("should collect errors from all codecs", () => {
  expectParseFailure(ab, {})
  expectParseFailure(ab, { a: "1", b: "1" })
})

test("should work with complex codecs", () => {
  const complex = intersection(ab, object({ c: fromJson.date }))
  const now = new Date()

  expectParseSuccess(
    complex,
    { a: 1, b: 1, c: now.toISOString() },
    { a: 1, b: 1, c: now }
  )
})
