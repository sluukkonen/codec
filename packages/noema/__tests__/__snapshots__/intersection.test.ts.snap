// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect errors from all codecs 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "path": [
      "a",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "path": [
      "b",
    ],
    "received": "undefined",
  },
]
`;

exports[`should collect errors from all codecs 2`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "path": [
      "a",
    ],
    "received": "string",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "path": [
      "b",
    ],
    "received": "string",
  },
]
`;

exports[`should fail to parse non-objects 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "path": [],
    "received": "null",
  },
]
`;
